front50-migrations/src/main/java/com/netflix/spinnaker/front50/migrations/MoveSubpathToLocationMigration.java

```diff
@@ -0,0 +1,132 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class MoveSubpathToLocationMigration implements Migration {
+
+  private static final Logger log = LoggerFactory.getLogger(CloudProvidersStringMigration.class);
+
+  // Only valid until October 1st, 2020
+  private static final Date VALID_UNTIL = new GregorianCalendar(2020, 10, 1).getTime();
+
+  private Clock clock = Clock.systemDefaultZone();
+
+  private final PipelineDAO pipelineDAO;
+
+  @Autowired
+  public MoveSubpathToLocationMigration(PipelineDAO pipelineDAO) {
+    this.pipelineDAO = pipelineDAO;
+  }
+
+  @Override
+  public boolean isValid() {
+    return clock.instant().toEpochMilli() < VALID_UNTIL.getTime();
+  }
+
+  @Override
+  public void run() {
+    log.info("Starting git/repo artifact migration");
+
+    Predicate<Pipeline> hasExpectedGitRepoArtifact =
+        p -> {
+          List<Map> expectedArtifacts = (List<Map>) p.get("expectedArtifacts");
+          if (expectedArtifacts == null) {
+            return false;
+          }
+
+          return expectedArtifacts.stream()
+              .anyMatch(
+                  e -> {
+                    Map matchArtifact = (Map<String, Object>) e.get("matchArtifact");
+                    Map defaultArtifact = (Map<String, Object>) e.get("defaultArtifact");
+
+                    return matchArtifact != null
+                            && isGitRepoArtifact(matchArtifact)
+                            && hasSubpathInMetadata(matchArtifact)
+                        || defaultArtifact != null
+                            && isGitRepoArtifact(defaultArtifact)
+                            && hasSubpathInMetadata(defaultArtifact);
+                  });
+        };
+
+    pipelineDAO.all().stream()
+        .filter(hasExpectedGitRepoArtifact)
+        .forEach(pipeline -> migrate(pipelineDAO, pipeline));
+  }
+
+  private boolean isGitRepoArtifact(Map artifact) {
+    return "git/repo".equalsIgnoreCase((String) artifact.get("type"));
+  }
+
+  private boolean hasSubpathInMetadata(Map artifact) {
+    Map<String, Object> metadata = (Map<String, Object>) artifact.get("metadata");
+    if (metadata != null) {
+      String location = Strings.nullToEmpty((String) artifact.get("location"));
+      String subpath = Strings.nullToEmpty((String) metadata.get("subPath"));
+      return !subpath.isEmpty() && location.isEmpty();
+    } else {
+      return false;
+    }
+  }
+
+  private void migrate(ItemDAO<Pipeline> dao, Pipeline pipeline) {
+
+    log.info(
+        "Move git/repo subpath to location field (application: {}, pipelineId: {}, expectedArtifacts: {})",
+        pipeline.getApplication(),
+        pipeline.getId(),
+        pipeline.get("expectedArtifacts"));
+
+    List<Map<String, Object>> expectedArtifacts =
+        (List<Map<String, Object>>) pipeline.get("expectedArtifacts");
+    for (Map<String, Object> expectedArtifact : expectedArtifacts) {
+      Map<String, Object> matchArtifact =
```

Comment:
Similar to the above comment, this would be more readable if there were a function
```void migrateArtifact(@Nullable Map<String, Object> artifact)```
so we don't have the logic repeated for match/expected.